### PR TITLE
Application attachments -> blobs

### PR DESCRIPTION
**Pull request by millerdev**
take two of #11308, now separated from blob migration refactoring so other PRs can use them too.

@dannyroberts @snopoke cc @czue @benrudolph

### CODE REVIEW

File: corehq/apps/app_manager/models.py [hunk 1]
```diff
@@ -4541,13 +4541,10 @@ def make_build(self, comment=None, user_id=None, previous_version=None):
...
-            # since this hard to put in a test
-            # I'm putting this assert here if copy._id is ever None
-            # which makes tests error
-            assert copy._id
-        except AssertionError:
-            raise
+        # since this hard to put in a test
+        # I'm putting this assert here if copy._id is ever None
+        # which makes tests error
+        assert copy._id
```
Comment (benrudolph) [inline corehq/apps/app_manager/models.py]: nice

File: corehq/apps/app_manager/models.py [hunk 2]
```diff
@@ -3856,38 +3876,92 @@ def wrap(cls, data):
                 self.lazy_put_attachment(name=name, **info)
         return self
 
-    def lazy_list_attachments(self):
-        keys = set()
-        keys.update(getattr(self, '_LAZY_ATTACHMENTS', None) or {})
-        keys.update(self.blobs or {})
-        return keys
+    def __attachment_cache_key(self, name):
```
Comment (benrudolph) [inline corehq/apps/app_manager/models.py]: out of curiosity why are you using `__` prefix

File: corehq/blobs/tests/test_migrate.py [hunk 3]
```diff
@@ -176,7 +176,7 @@ def test_migrate_saved_exports(self):
         # setup data
         app = Application()
         app.save()
-        payload = '<fake xform source />'
+        payload = u'<fake xform source>\u2713</fake>'
```
Comment (benrudolph) [inline corehq/blobs/tests/test_migrate.py]: ✔️

**Comment by benrudolph** [comment]
lg2m, will let someone with more context look it over

**Comment by dannyroberts** [comment]
👍 looks good, @millerdev let me know how you want to move forward!

**Comment by millerdev** [comment]
I've had this on staging (and ran the migration there) since last week and I have not seen or heard of any issues.  I'll probably roll it out to production stacks sometime this week.

**Comment by millerdev** [comment]
Rebased on master. Ready for final review/merge.

**Comment by millerdev** [comment]
(wait for tests)

**Comment by millerdev** [comment]
✅

File: corehq/apps/app_manager/models.py [hunk 4]
```diff
@@ -3845,7 +3848,88 @@ def validate_for_build(self):
...
+                pre_save()
+
+            def del_pre_save():
+                del self._PRE_SAVE
+
+            self.register_post_save(del_pre_save)
+
+        if self._LAZY_ATTACHMENTS and self.__save_params is None:
+            assert self.__save_params is None, (self.__save_params, params)
+            self.__save_params = params
```
Comment (snopoke) [inline corehq/apps/app_manager/models.py]: nit: This doesn't seem to get used inside this `if` branch and gets deleted before exiting the branch. Is it used by `self.atomic_blobs`?

Comment (snopoke) [inline corehq/apps/app_manager/models.py]: removed in later commit

File: corehq/blobs/tests/test_migrate.py [hunk 5]
```diff
@@ -167,6 +168,53 @@ def modify():
...
+        def modify():
+            doc = Application.get(app._id)
+            doc.put_attachment(new_payload, "form.xml")  # calls save()
+
+        self.do_failed_migration({app: (1, 1)}, modify)
+
+        # verify: attachments were not migrated
+        exp = Application.get(app._id)
+        self.assertEqual(exp.fetch_attachment("form.xml"), new_payload)
+        self.assertEqual(exp.fetch_attachment("other.xml"), old_payload)
```
Comment (snopoke) [inline corehq/blobs/tests/test_migrate.py]: not clear how this verifies that the migrations wasn't done

**Pull request merged**